riscv/csrs.cc

```diff
@@ -1400,3 +1412,39 @@ void henvcfg_csr_t::verify_permissions(insn_t insn, bool write) const {
       throw trap_illegal_instruction(insn.bits());
   }
 }
+
+bool henvcfg_csr_t::unlogged_write(const reg_t val) noexcept {
+  // mask out read-only 0 bits if menvcfg has them turned off
+  return masked_csr_t::unlogged_write((menvcfg->read() | ~(MENVCFG_PBMTE | MENVCFG_STCE)) & val);
+}
```

Comment:
This does not seem necessary when you are also masking on reads.

Reply:
I'm being pedantic here because writes to henvcfg would retain values to those bits when the corresponding bits are off in menvcfg and later turned on (not technically read-only).

Reply:
If no CSRW is performed while `menvcfg` is masking these bits, the old values will be revealed when `menvcfg` re-enables these bits. If any CSRW happens, this code will 0 those two bits. That is pretty bizarre behavior.

I'm not sure what the spec requires in this case, but I'm pretty sure it doesn't require this behavior. If it's like other bits that are read-only 0 and later become writable, the bits become undefined at the time they become writable. (For example, hypervisor-only state [such as `mstatus.GVA`] when `misa.H` goes from 0 to 1.)